### DISCUSSION TITLE
Custom MIDI mapping integration for each instrument

### DISCUSSION THREAD

**Discussion by AndreasL50** [hydrogen-music/hydrogen#1757] [General]
In the FAQ, it literally states, that Hydrogen is not supporting MIDI mapping, even though the documentation does, but won't tell you how. That leads me to be confused. I'm surprised I haven't found much regarding this, which makes me even more confused. I think this is one of the most essential features for a DAW with MIDI integration.

After hours, I couldn't find anything in that regard, except for external means to achieve that.
Without that, the whole application is utterly useless to me.
One would assume, that a DAW lets you actually use an electronic instrument, but that would be too useful for any normal use case, I suppose.
Not getting any hints/indications/information about this, the intuitive and usability goal has been missed greatly in my opinion.

MIDI is coming through, but a connected TD-17 won't work, because Hydrogen won't let me define what the MIDI signals mean and it can't figure it out on its own. The out of the box settings should at least give me kick, snare, hi-hat, but not even that. I also have no way of actually finding out what hard coded the mapping is, though I did find one in the documentation, which makes me unsure if this even is the problem, but other DAWs have no problems, so that the fact it can read MIDI correctly, just tells me the mapping is off. The sounds itself work fine too.
All I can see is the response through the small indicator on top and the fact it can read correct MIDI when setting up events/actions.
It should be as simple as pressing a button, hitting the instrument / sending a MIDI signal and finding that it indeed works. That should be one of the first things to exist for any DAW with MIDI integration.

Do I really need an external controller to reroute the MIDI beforehand? If this really is the case, then this is quite ridiculous.
Did I miss or misunderstand anything? Where can I find sources that indeed explain what to do?
I ask that, because the Hydrogen manual sometimes just tells you something is possible, without telling you anything else, which is no help at all.

**Comment by theGreatWhiteShark** [comment]
Hey @AndreasLang50 ,

Hydrogen has two ways of dealing with incoming MIDI notes
1. It maps all instruments of your drumkit to incoming MIDI note keys (regardless of the channel used) starting at note key 36. When choosing this option you can control mapping by drag-dropping the instruments into your favorite order.
2. When enabling 'Use output note as input note' in Preferences > MIDI System, you can control both MIDI key and channel of each instrument in the instrument editor http://hydrogen-music.org/documentation/manual/manual_en.html#chpt.instrument_editor.general .

If you are unsure which keys are sent you can open the log and set log level to "Info". All incoming notes should now be listed.

> The out of the box settings should at least give me kick, snare, hi-hat

There is no general truth regarding which MIDI note corresponds to part of a drumkit. There is [General MIDI](https://en.wikipedia.org/wiki/General_MIDI#Percussion) but this only fits very small kits. So, we let those people who create the drumkits decide about the mapping by defining both the order of instruments as well as the output note keys for each of them.

**Reply by AndreasL50** [reply]
I find this a bit unintuitive, but indeed that does make sense.

The exact section of the documentation you linked still confuses me.
"Hydrogen is capable of generating MIDI messages that you can use to trigger any external MIDI device or application."
"From now on every time a note is played for that instrument (in the Hydrogen sequencer) a MIDI message will be sent to your external app/device"
That doesn't sound like a "MIDI input, sound triggering issue WITHIN hydrogen" issue to me at all. That's a completely different topic, so it shouldn't be linked this greatly IMO.

It already had most of the instruments correctly linked (as I suspected), but for some reason, the checkbox option "Disregard MIDI messages after MIDI-Action took place" under Preferences > MIDI System had no effect when changing the value.
After deleting the hydrogen.conf file and unchecking the checkbox, it worked.
For me, "Ignore Note-Off" also has to be checked, or else it'll constantly choke.

Thanks for your help, the log helped me out.
But my opinion about how to configure this remains more or less the same (simply because it cost me a lot of time and I got frustrated, because I found absolutely nothing even remotely related to this very simple issue that seemed appropriate. Getting confused by the documentation just added fuel).

Something like a little hint popping up when no sounds are played, but MIDI is steadily coming through, would have been enough.

**Reply by theGreatWhiteShark** [reply]
> I find this a bit unintuitive

True. But Hydrogen is an old application (more than 22 years) and lots of people are accustomed to the way it works. We are just here to maintain it and to smooth its UX without introducing breaking changes.

> Something like a little hint popping up when no sounds are played, but MIDI is steadily coming through, would have been enough.

Well, depending on your workflow this could be quite annoying. E.g. you could have several programs connected to your MIDI output and only handle some parts messages by Hydrogen itself.

> It already had most of the instruments correctly linked (as I suspected), but for some reason, the checkbox option "Disregard MIDI messages after MIDI-Action took place" under Preferences > MIDI System had no effect when changing the value.
After deleting the hydrogen.conf file and unchecking the checkbox, it worked.

That's odd. I will have a look.

> For me, "Ignore Note-Off" also has to be checked, or else it'll constantly choke.

That's because you MIDI source is choking the notes by sending NOTEON followed by NOTEOFF messages. Anyway, I disable it on my MIDI keyboard too.

Thanks for the hints regarding the documentation. I rewrote the whole thing recently because it was quite out of date. It is still rough around the edges and can use some polishing.

**Reply by AndreasL50** [reply]
> Well, depending on your workflow this could be quite annoying
A simple "Do not show again" would be enough, right?

> That's odd. I will have a look
It could have been a permission problem, as I had an accident before

> Thanks for the hints regarding the documentation. I rewrote the whole thing recently because it was quite out of date. It is still rough around the edges and can use some polishing.

I don't exactly know where it was stated, but I read that Hydrogen strives to be used intuitively and that triggered me in that moment.
Though, the look is definitely is very good for 22 years. Good job on that.

I'm having a few issues setting up my hi-hat levels properly, but I should be able to work it out. Thanks for your help

**Reply by theGreatWhiteShark** [reply]
>  >  It already had most of the instruments correctly linked (as I suspected), but for some reason, the checkbox option "Disregard MIDI messages after MIDI-Action took place" under Preferences > MIDI System had no effect when changing the value.
    After deleting the hydrogen.conf file and unchecking the checkbox, it worked.
>
> That's odd. I will have a look.

That turned out to be a bug and should be fixed now in version 1.2.1.

I also took your input to heart and created a dedicated [MIDI chapter](http://hydrogen-music.org/documentation/manual/manual_en.html#chpt.midi) within the manual explaining thoroughly how to set up input and output connections.
